### PR TITLE
Add possibility to apply latest MET Phi corrections during MiniAOD processing

### PR DESCRIPTION
**Pull request by michaelwassmer**
#### PR description:

This PR is intended to allow the application of the latest MET Phi corrections when running over MiniAOD. The MET Phi corrections are intended to mitigate the observed azimuthal angle modulation of missing transverse momentum (MET). The corrections were derived by @lathomas and can be found at the following link: https://lathomas.web.cern.ch/lathomas/METStuff/XYCorrections/
This PR is just the CMSSW implementation of those corrections using an already existing module with some adaptions.
The effect of the PR can be seen in more detail in the following two presentations, which were given in the MET meeting:
1. https://indico.cern.ch/event/1153621/contributions/4849172/attachments/2431720/4164278/presentation.pdf
2. https://indico.cern.ch/event/1166184/contributions/4897784/attachments/2453553/4204792/presentation.pdf

**Important:** The PR is designed such that it shouldn't affect the default/current behavior at all.
For the PR to actually do something, the newly added 'campaign' and 'era' input arguments of the MET tool need to be explicitly used. This is actually one thing, where the input of experts would be greatly appreciated. It would probably be preferred to automatically retrieve the campaign and era, but at the moment I'm not sure how this could be achieved.

Changes:
1. JetMETCorrections/Type1MET/plugins/MultShiftMETcorrInputProducer.cc
- Added possibility to also use the total number of primary vertices as the variable the Phi corrections depend on
- Removed the PF candidate loop if the corrections are done as a function of a variable which doesn't use the PF candidates
2. Added several python configuration files containing the MET Phi corrections
3. PhysicsTools/PatUtils/python/patPFMETCorrections_cff.py
- Added the corrections from the configuration files as objects to be able to use them
4. PhysicsTools/PatUtils/python/tools/runMETCorrectionsAndUncertainties.py
- Allow to use the new corrections in the MET module if the newly added 'era' and 'campaign' input arguments are explicitly used
- The default behavior should stay untouched for now
 

#### PR validation:

Successfully performed  `code-checks` and `code-format`.
Besides some file read/open errors, the `runTheMatrix.py -l limited -i all` command also seemed to work successfully.

### CODE REVIEW

File: PhysicsTools/PatUtils/python/tools/runMETCorrectionsAndUncertainties.py [hunk 1]
```diff
@@ -1072,33 +1084,24 @@ def removePostfix(self, name, postfix):
...
-            "T1T2Txy_25ns":metCors.patMultPhiCorrParams_T1T2Txy_25ns,
-            "T0pcT1T2Txy_25ns":metCors.patMultPhiCorrParams_T0pcT1T2Txy_25ns,
-            "T1SmearTxy_25ns":metCors.patMultPhiCorrParams_T1SmearTxy_25ns,
-            "T1T2SmearTxy_25ns":metCors.patMultPhiCorrParams_T1T2SmearTxy_25ns,
-            "T0pcT1SmearTxy_25ns":metCors.patMultPhiCorrParams_T0pcT1SmearTxy_25ns,
-            "T0pcT1T2SmearTxy_25ns":metCors.patMultPhiCorrParams_T0pcT1T2SmearTxy_25ns
-            }
-
-        getattr(process, "patPFMetTxyCorr"+postfix).parameters = xyTags[corScheme+"_25ns"]
+        print("DEBUG ",campaign,datamc,era)
```
Comment (jpata) [inline PhysicsTools/PatUtils/python/tools/runMETCorrectionsAndUncertainties.py:1090]: unintentionally left in?

**Review by jpata — COMMENTED** [review]
> This is actually one thing, where the input of experts would be greatly appreciated. It would probably be preferred to automatically retrieve the campaign and era, but at the moment I'm not sure how this could be achieved.

Don't we usually configure this on the fly with the era-dependent `toModify` operations?
https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L38

File: PhysicsTools/PatUtils/python/tools/runMETCorrectionsAndUncertainties.py (hunk 1 above)
Comment (michaelwassmer) [reply PhysicsTools/PatUtils/python/tools/runMETCorrectionsAndUncertainties.py]: Yes, I forgot that line. I just removed it and updated the PR.
Regarding the modifiers, are they also capable of differentiating between 'suberas' e.g. 2018B and 2018C? The corrections also differ between the these suberas.

**Comment by jpata** [comment]
> Regarding the modifiers, are they also capable of differentiating between 'suberas' e.g. 2018B and 2018C? The corrections also differ between the these suberas.

Ah, I see. As far as I understand, the cmsRun python configuration is not at this level of granularity. Is this something that should be coming from the GT instead @cms-sw/alca-l2.

**Comment by tvami** [comment]
> Is this something that should be coming from the GT instead https://github.com/orgs/cms-sw/teams/alca-l2.

I confirm that this is what conditions are meant for.
I didnt read this PR (yet) but this would probably mean an introduction of a new CondFormat which is now phi dependent.

**Comment by jpata** [comment]
hold

until the possible interaction with GT is figured out

**Comment by michaelwassmer** [comment]
All right, just some comments/thoughts of the possibilities we then have:
1. Provide this as it is currently so that people who are still using MiniAOD frameworks can apply these corrections on the fly by calling the MET module and then reading the new/corrected MET from the event. Not sure how many people still use MiniAOD frameworks but should make their lives easier if they desire the MET phi corrections.
2. For the official production, I guess there is not much other possibility than putting this into a GT if the corrections would be desired for the official production.

**Comment by jpata** [comment]
Having a centrally available correction (option 2) would be preferred to a custom MiniAOD recipe.

**Comment by tvami** [comment]
@michaelwassmer let me understand the final goal here, is this supposed to be backported to 10_6_X so it's used for UL ?

**Comment by michaelwassmer** [comment]
> @michaelwassmer let me understand the final goal here, is this supposed to be backported to 10_6_X so it's used for UL ?

This could be one use case to have phi corrected MET in the next MiniAOD interation out of the box. However, I asked the JME/MET conveners to join on Friday so we have their perspective as well.

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
+reconstruction
- possibility to enable MET phi corrections in private MINIAOD production
- not enabled by default
- a follow-up for enabling this via the global tag is necessary

**Comment by clacaputo** [comment]
Can `hold` be removed only by the person that triggers it?

**Comment by perrotta** [comment]
unhold

**Comment by qliphy** [comment]
please test
to refresh the results

**Comment by perrotta** [comment]
+1
- Intended for private analyses and productions

**Pull request merged**